### ISSUE TITLE
Proposed expected to pass, expected to fail workflows

### ISSUE THREAD

**Issue by nurturenature** [canonical/jepsen.dqlite#94]
This issue proposes splitting the single `Dqlite Jepsen tests` workflow into two separate workflows: expected to pass and expected to fail.

- With current known failures, e.g. pause, one always has to look into the workflow run and examine the failed jobs to see if there's anything new or interesting.
- The single workflow failing doesn't reflect that many/most tests are passing.
- Could more easily evaluate commits, e.g. do changes fix and/or break the status quo.

Expected to pass:
- existing `test.yml`
- name: Dqlite Jepsen tests (expected pass)
- run
  - every hour
  - on pull_request, push
  - on_dispatch (manually)

Expected to fail:
- new `test-fail.yml`
- name: Dqlite Jepsen tests (expected fail)
- run
  - every hour
  - on_dispatch (manually)

Callable workflow:
- new `test-build-run.yml`
- more configurable version of `test` job steps currently in `test.yml`
- only `on_call`, as if a GitHub yaml function

----

Implementation:

- `test.yml`
  ```yaml
  name: Dqlite Jepsen tests (expected pass)
  
  on:
    push:
    pull_request:
    schedule:
      - cron: "0 * * * *"
    workflow_dispatch:
      ...

  jobs:
    expected-pass:
      uses: canonical/jepsen.dqlite/.github/workflows/test-build-run.yml@master
      with:
        nemesis : >
          [ "none", "stop", "kill", "partition", "disk", "member",
            "stop,partition", "kill,partition", "member,partition" ]
  ```

- `test-fail.yml`
  ```yaml
  name: Dqlite Jepsen tests (expected fail)
  
  on:
    schedule:
      - cron: "30 * * * *"
    workflow_dispatch:
      ...
  
  jobs:
    expected-fail:
      uses: canonical/jepsen.dqlite/.github/workflows/test-build-run.yml@master
      with:
        nemesis : >
          [ "pause", "pause,partition"
            "disk,partition", "disk,pause" ]
  ```

- `test-build-run.yml`
  ```yaml
  on:
    workflow_call:
      inputs:
        nemesis:
          type: string
          required: true
  
  jobs:
    test:
      strategy:
        matrix:
          workload: [ append, set, bank ]
          nemesis: ${{ fromJSON(inputs.nemesis) }}
      steps:
      - name: Checkout code
      ...
  ```

----

It could also enable future expansion of test run configurations.

```yaml
jobs:
  custom-run:
    uses: canonical/jepsen.dqlite/.github/workflows/test-build-run.yml@master
    with:
      workload: [ bank ]
      nemesis : [ disk ]
      opts: >
        [ "--node-targets primaries"
          "--test-count 10" ]
```

----

Would this be a more helpful test workflow?

**Comment by nurturenature** [comment]
> configurations that only sometimes cause a known bug -- should those go in test.yml or test-fail.yml?

If it's known, test-fail.

> This is the workflow that would trigger when we merge something in raft or dqlite, right? i.e. https://github.com/canonical/jepsen.dqlite/issues/43.

Yes.
A generic way to call and run arbitrary test configs.

----

> ask what your dev setup for hacking on jepsen.dqlite looks like?

I use my local host as the control node with LXC for db nodes. Jepsen is [developed](https://github.com/jepsen-io/jepsen#lxc) and tested with Debian LXC.

> I wrote [this script](https://github.com/canonical/jepsen.dqlite/blob/master/test.sh)

That's good. It's the most minimal way to build a Jepesen environment that I've seen.

**Comment by nurturenature** [comment]
Closed by #97.
